example/main.go

```diff
@@ -17,23 +17,39 @@
 package main
 
 import (
+	"context"
+
 	"sigs.k8s.io/kube-scheduler-wasm-extension/guest"
-	"sigs.k8s.io/kube-scheduler-wasm-extension/guest/api"
+	"sigs.k8s.io/kube-scheduler-wasm-extension/guest/api/framework"
 )
 
 func main() {
-	guest.Filter = api.FilterFunc(nameEqualsPodSpec)
+	guest.RegisterFilter(&NodeName{})
+}
+
+// NodeName is a plugin that checks if a pod spec node name matches the current node.
+type NodeName struct{}
+
+const (
+	// Name is the name of the plugin used in the plugin registry and configurations.
+	name = "NodeName"
+
+	// ErrReason returned when node name doesn't match.
+	ErrReason = "node(s) didn't match the requested node name"
+)
+
+func (*NodeName) Name() string {
+	return name
 }
 
-// nameEqualsPodSpec schedules this node if its name equals its pod spec.
-func nameEqualsPodSpec(args api.FilterArgs) (api.StatusCode, string) {
-	nodeName := nilToEmpty(args.NodeInfo().Node().Metadata.Name)
-	podSpecNodeName := nilToEmpty(args.Pod().Spec.NodeName)
+func (*NodeName) Filter(ctx context.Context, state *framework.CycleState, podFn framework.PodFn, nodeInfoFn framework.NodeInfoFn) *framework.Status {
+	nodeName := nilToEmpty(nodeInfoFn().Node().Metadata.Name)
+	podSpecNodeName := nilToEmpty(podFn().Spec.NodeName)
 
 	if len(podSpecNodeName) == 0 || podSpecNodeName == nodeName {
-		return api.StatusCodeSuccess, ""
+		return nil
```

Comment:
I think we are trying too hard to be exactly the same as the host code, and this won't be the same. Not only is the context parameter questionable, but so is this, We are adding the question of "why and when should we return nil?" vs passing the two components.

Reply:
In the scheduler plugins, nil means success.
https://github.com/kubernetes/kubernetes/blob/bdf34b3f56d269d9b435a882145aa7fe217691ce/pkg/scheduler/framework/interface.go#L78

Reply:
I'm fine with going conventionally. Just remember this plugin will eventually support other languages and also not necessarily be slow because proto reasons. We can proceed to optimize for conventionality instead of performance, just remember performance can degrade quickly and some people will fault the plugin after it reaches some threshold. If we ignore performance over a long time it can be difficult to recover.